File: javascript/functional-js/type-check.js
// type_check(string, function) -> function -> value or type error
// Wraps single-argument function in a function that checks the type of its argument.
// const add1 = type_check("number", (value) => value + 1) -> add1(10) -> 11

function type_check(type, action) {
    return function (value) {
        if (typeof value !== type) {
            throw new TypeError("Argument must be of type " + type);
        }
        return action(value);
    };
}

File: python/functional-py/type-check.py
# type_check(type, *function) -> *function
# Inputs a type and variadic function and returns 
# a variadic function that checks the types of its inputs.

# def add(*numbers):
#     sum = 0
#     for number in numbers:
#         sum += number
#     return sum

# add = type_check(int, add)
# add(1, 6) -> 7

def type_check(check, action):
    def wrapper(*inputs):
        for input in inputs:
            if type(input) != check:
                raise TypeError(f"{input} must be of type {check}")
        return action(*inputs)
    return wrapper
